app/models/forms/integrity_warnings/warner.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Forms
+  module IntegrityWarnings
+    # Enumerates integrity warnings for Forms
+    class Warner
+      MAX_FORMS = 3
+
+      attr_accessor :object
+
+      def initialize(object)
+        self.object = object
+      end
+
+      def warnings(type)
+        # Expects the subclass method with name `type` to return an array of
+        # symbols or 1-element hashes of form [symbol => symbol].
+        # The first symbol is the boolean method to call on the object
+        # to check if the warning should be shown.
+        # The second symbol (in the array case) is a method to be called locally to provide extra
+        # info for I18n.
+        send(type).map do |params|
```

Comment:
Love the abstraction, but can this be made more explicit? E.g. named args `should_warn` and `i18n_info` instead of arrays of implicitly different symbols.